### PR TITLE
[dbnode] Add metric to count times when timeseries annotation changes

### PR DESCRIPTION
**Pull request by vpranckaitis**
**What this PR does / why we need it**:

Adds a metric to count time when annotation of a timeseries change. This information is useful since constantly changing annotations require additional storage. 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/dbnode/encoding/m3tsz/timestamp_encoder.go [hunk 1]
```diff
@@ -182,6 +194,9 @@ func (enc *TimestampEncoder) writeAnnotation(stream encoding.OStream, ant ts.Ann
 	stream.WriteBytes(buf[:annotationLength])
 	stream.WriteBytes(ant)
 
+	if enc.PrevAnnotationChecksum != emptyAnnotationChecksum {
+		enc.metrics.annotationRewritten.Inc(1)
```
Comment (linasm) [inline src/dbnode/encoding/m3tsz/timestamp_encoder.go:198]: Just add a comment (as we discussed offline) to explain that this is based on the expectation that normally the annotation would not get rewritten, and might be too expensive to keep this around if this expectation turns out to be false someday.

Comment (vpranckaitis) [reply src/dbnode/encoding/m3tsz/timestamp_encoder.go]: https://github.com/m3db/m3/pull/3565/commits/61f0826830f3386f4b2d70344472c29496d5be36

File: src/dbnode/encoding/m3tsz/timestamp_encoder.go [hunk 2]
```diff
@@ -182,6 +185,12 @@ func (enc *TimestampEncoder) writeAnnotation(stream encoding.OStream, ant ts.Ann
 	stream.WriteBytes(buf[:annotationLength])
 	stream.WriteBytes(ant)
 
+	if enc.PrevAnnotationChecksum != emptyAnnotationChecksum {
+		// NB: current assumption is that in the most cases no annotation writes should happen and that
```
Comment (linasm) [inline src/dbnode/encoding/m3tsz/timestamp_encoder.go:189]: `no annotation writes should happen` I think is a bit misleading - we expect to have one annotation per block on every time series. I think the comment should speak of rewrites specifically, and not writes in general.

File: src/dbnode/encoding/options.go [hunk 3]
```diff
@@ -61,6 +63,7 @@ func newOptions() Options {
 		byteFieldDictLRUSize:   defaultByteFieldDictLRUSize,
 		iStreamReaderSizeM3TSZ: defaultIStreamReaderSizeM3TSZ,
 		iStreamReaderSizeProto: defaultIStreamReaderSizeProto,
+		metrics:                NewMetrics(instrument.NewOptions().MetricsScope()),
```
Comment (linasm) [inline src/dbnode/encoding/options.go:66]: Not really sure if we should be creating new `instrument.Options` here, disconnected from the globally used one.

File: src/dbnode/storage/block/options.go [hunk 4]
```diff
@@ -59,6 +60,7 @@ type options struct {
 
 // NewOptions creates new database block options
 func NewOptions() Options {
+	iOpts := instrument.NewOptions()
```
Comment (linasm) [inline src/dbnode/storage/block/options.go:63]: Same here.

File: src/dbnode/encoding/m3tsz/timestamp_encoder.go (hunk 2 above)
Comment (vpranckaitis) [reply src/dbnode/encoding/m3tsz/timestamp_encoder.go]: https://github.com/m3db/m3/pull/3565/commits/3833ce9083a885a585ab43f0652e06bdbc03a8d7

File: src/dbnode/encoding/options.go (hunk 3 above)
Comment (vpranckaitis) [reply src/dbnode/encoding/options.go:66]: This is done in other places, too. E.g. 

https://github.com/m3db/m3/blob/a2a70d0168c272415bb914a6aaab1e803b7d77cf/src/dbnode/storage/options.go#L208-L213

https://github.com/m3db/m3/blob/a2a70d0168c272415bb914a6aaab1e803b7d77cf/src/dbnode/storage/bootstrap/bootstrapper/fs/options.go#L81-L83

To me, it seems that the pattern is to initialize options with some sensible defaults and then inject dependencies using methods (which sometimes is not very convenient).

**Review by linasm — APPROVED** [review]
LGTM

File: src/dbnode/encoding/metrics.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// TimestampEncoderMetrics contains timestamp encoder metrics.
+type TimestampEncoderMetrics struct {
+	annotationRewritten tally.Counter
+}
+
+func newTimestampEncoderMetrics(scope tally.Scope) TimestampEncoderMetrics {
+	return TimestampEncoderMetrics{
+		annotationRewritten: scope.Counter("annotation-rewritten"),
```
Comment (vpranckaitis) [inline src/dbnode/encoding/metrics.go:44]: The resulting metric name is `timestamp_encoder_annotation_rewritten`

**Review by robskillington — APPROVED** [review]
LGTM

**Pull request merged**